packages/contracts-bedrock/contracts/L1/L2OutputOracle.sol

```diff
@@ -143,34 +128,26 @@ contract L2OutputOracle is OwnableUpgradeable, Semver {
 
     /**
      * @notice Deletes all output proposals after and including the proposal that corresponds to
-     *         the given block number. Can only be called by the owner, but will be replaced with
+     *         the given output index. Can only be called by the owner, but will be replaced with
      *         a mechanism that allows a challenger contract to delete proposals.
      *
-     * @param _l2BlockNumber L2 block number of the first output root to delete.
+     * @param _l2OutputIndex Index of the first L2 output to be deleted. All outputs after this
+     *                       output will also be deleted.
      */
     // solhint-disable-next-line ordering
-    function deleteL2Outputs(uint256 _l2BlockNumber) external onlyOwner {
-        // Simple check that accomplishes two things:
-        //   1. Prevents deleting anything before (and including) the starting block.
-        //   2. Prevents deleting anything other than a checkpoint block.
+    function deleteL2Outputs(uint256 _l2OutputIndex) external onlyOwner {
+        // Make sure we're not *increasing* the length of the array.
         require(
-            l2Outputs[_l2BlockNumber].outputRoot != bytes32(0),
-            "L2OutputOracle: cannot delete a non-existent output"
+            _l2OutputIndex < l2Outputs.length,
+            "L2OutputOracle: cannot delete outputs after the latest output index"
         );
 
-        // Prevent deleting beyond latest block number. Above check will miss this case if we
-        // already deleted an output and then the user tries to delete a later output.
-        require(
-            _l2BlockNumber <= latestBlockNumber,
-            "L2OutputOracle: cannot delete outputs after the latest block number"
-        );
-
-        // We're setting the latest block number back to the checkpoint block before the given L2
-        // block number. Next proposal will overwrite the deleted output and following proposals
-        // will delete any outputs after that.
-        latestBlockNumber = _l2BlockNumber - SUBMISSION_INTERVAL;
+        // Use assembly to delete the array elements because Solidity doesn't allow it.
+        assembly {
+            sstore(l2Outputs.slot, _l2OutputIndex)
+        }
```

Comment:
Wondering if there is an off by 1 error here

Reply:
I think it's correct. Imagine if there are 2 elements in the array. If we want to delete the second element (element index 1) then we give `_l2OutputIndex = 1`. So we store `1` as the length and then the array becomes 1 element long, which deletes element 2.